toolchain/driver/runtimes_cache.h

```diff
@@ -0,0 +1,414 @@
+  //
+  // Note, nothing goes _wrong_ if this deadline is exceeded, but multiple
+  // copies of runtimes may end up being built and all but one thrown away.
+  static constexpr Filesystem::Duration BuildLockDeadline =
+      std::chrono::seconds(200);
+  // The interval at which to poll for a build lock. This needs to be small
+  // enough that we don't waste an excessive amount of time if a build of the
+  // runtimes completes *just* after a poll. Typically, that means we want this
+  // to be significant lower than the expected time it would take to build the
+  // runtimes. Note that we don't poll if the runtimes have been completely
+  // built prior to the query coming in, so this doesn't form the _minimum_ time
+  // to find a component of the runtimes tree.
+  static constexpr Filesystem::Duration BuildLockPollInterval =
+      std::chrono::milliseconds(200);
+
+  // The path to the clang resource directory within the runtimes.
+  //
+  // This uses `std::string_view` to simply using with paths.
+  static constexpr auto ComponentPath(Component component) -> std::string_view {
+    switch (component) {
+      case ClangResourceDir:
+        return "clang_resource_dir";
+      case NumComponents:
+        CARBON_FATAL("Invalid component");
+    }
+  }
+
+  // A format string used to form the lock file for a given directory in the
+  // runtimes. This needs to be C-string, so directly expose the character
+  // array.
+  static constexpr char LockFileFormat[] = ".{0}.lock";
+
+  explicit Runtimes(std::filesystem::path base_path, Filesystem::Dir base_dir,
+                    Filesystem::WriteFile lock_file, Filesystem::FileLock flock,
+                    llvm::raw_ostream* vlog_stream = nullptr)
+      : base_path_(std::move(base_path)),
+        base_dir_(std::move(base_dir)),
+        lock_file_(std::move(lock_file)),
+        flock_(std::move(flock)),
+        vlog_stream_(vlog_stream) {
+    CARBON_CHECK(base_path_.is_absolute(),
+                 "The base path must be absolute: {0}", base_path_);
+  }
+
+  // Implementation of building the Clang resource directory. This exposes the
+  // deadline and poll interval to allow testing with artificial values.
+  auto BuildImpl(Component component, Filesystem::Duration deadline,
+                 Filesystem::Duration poll_interval)
+      -> ErrorOr<std::variant<std::filesystem::path, Builder>>;
+
+  auto Destroy() -> void;
+
+  std::filesystem::path base_path_;
+  Filesystem::Dir base_dir_;
+  Filesystem::WriteFile lock_file_;
+  Filesystem::FileLock flock_;
+  llvm::raw_ostream* vlog_stream_ = nullptr;
+};
+
+// A managed cache of `Runtimes` directories.
+//
+// This class manages and provides access to a cache of runtimes. Each entry in
+// the cache is a collection of runtimes for a specific set of `Feature`s in a
+// directory that can be modeled with an object of the `Runtimes` type. We call
+// these entries "runtimes" typically. The cache keys and looks up these
```

Comment:
```suggestion
// This class manages and provides access to a cache of runtimes. Each entry in
// the cache is a runtimes directory for a specific set of `Feature`s, represented by an
// object of the `Runtimes` type. The word "runtimes" can refer to a single entry or
// to multiple entries, depending on the context. The cache keys and looks up these
```

Two separate changes here, both optional:
- Using the previously-defined term "runtimes directory" rather than re-describing the thing it refers to. This is both for brevity and because avoiding that term gives the impression that it's somehow not applicable.
- Expanding the terminology clarification to be more explicit about the two meanings of "runtimes".

Reply:
The first part of this makes sense to me, but I think the second part isn't quite right...

The word "runtimes" should always refer to a single entry in the cache. The *contents* of that entry might be a single component or multiple components, but I think we always want to think of an entry as a collection of (potentially) multiple runtime components, which is why it is always plural. I'll make a change to somewhat emphasize this?

Reply:
OK, that change is helpful. I've flagged a few places where "runtimes" seems to refer to multiple cache entries, now that I know that's not intentional.